### PR TITLE
Added a return type for Instagram generator method

### PR DESCRIPTION
**Pull request by goooseman**
The return type of `Hashtag`, `Location`, `Post`, `User` classes' generator function was missing. 

It is a problem to use it outside of the project, because you need to console.log the result first, than to create a type from it.

So it is better to write this definition in the project itself.

Looking forward for the code review!

### CODE REVIEW

**Comment by goooseman** [comment]
I totally agree with you. 

So the plan will be the following:

- [ ] Create a runtime type checker, that will check the response from the API
- [ ] Implement it inside Instagram class. If strict mode is enabled, it will throw an error, if not, it will send a winston log with `level: warning`
- [ ] Implement a test, that runs in a strict mode and checks, that no errors has been thrown
- [ ] Create a [cron job in Travis CI](https://docs.travis-ci.com/user/cron-jobs/) to run this test on a daily basis

The only issue right now, is that typescript removes all the type information on build time so there are no direct options to typecheck objects on runtime using Typescript interface.

I will speak with my colleges tomorrow about this issue. For now the plan A is to rewrite the interface with [io-ts](https://medium.com/@ottoki/runtime-type-checking-with-io-ts-in-typescript-14465169fb02).

After this PR will be finished, I will come back to rest of them.

**Comment by ScriptSmith** [comment]
Great, io-ts seems like a good option. The build has already been running as a daily cron job, so you can tick that one off.

**Comment by goooseman** [comment]
Ok, I've finished

What've been done:

1. A new util has been created to get an actual Instagram API response for Posts and SinglePosts (SinglePost is the result of scraping single Instagram post)
1. A new util has been created to generate actual `io-ts` types from the JSON. It supposed to work fully automatic, but the result is not perfect at all. So `README` has been written to describe what is going on.
1. Implemented a generic for `Instagram`, so the return type (`Post` or `SinglePost` can be specified)
1. Implemented a runtime check, so a warning is writing to logs, if types are not the same
1. Written a test to make sure, that warnings are written
1. Implemented a `strict` flag. If it is true, the Error will be thrown when validations failed
1. Written a test to make sure, that error is thrown

**Comment by goooseman** [comment]
The tests are failing, checking out

**Comment by goooseman** [comment]
Fixed all the typings (I hope).

But right now there is a new problem. The `Post` class fails to evaluate data from response (`instagram.ts:264`):

Screen Shot 2019-07-12 at 0 50 14

And I can see the following error:

Screen Shot 2019-07-12 at 0 50 40

Will check tomorrow, but maybe help is needed to understand how to query the result.

**Comment by goooseman** [comment]
Ah no, these are my type validations are failing. Will check tomorrow.

**Comment by goooseman** [comment]
It is now ready for a review.

The tests are failing, only on lts version of node. Strange, on my computer they are passed. Will check tomorrow what is the issue.

But anyway, waiting for the code review)

**Comment by goooseman** [comment]
Ok, ready for the review

**Comment by goooseman** [comment]
What has been done:

1. Tests for Full mode are added. 
2. Typings for Full mode is written. They are the same as `SinglePost`, but with the `location` as an `object`, not a `string`.
3. Different validators are passed for full mode and not a full mode.
1. New static type `TFullApiPost` is passed as generic to `Instagram` class from `Hashtag`, `User`, `Location`.
4. Types are renamed to avoid these `Node1`, `Node2`.
5. The issue about closing `postPage` in `instagram.ts` is fixed.
6. New FAQ section is written to redirect users in the warning message
7. A new instruction abouе fixing the typings is written.

**Comment by goooseman** [comment]
As far as I understand, only the owner of the repository can create an [issue template](https://help.github.com/en/articles/creating-issue-templates-for-your-repository)

**Comment by goooseman** [comment]
About difficulty of the reacting to API changes.

### Minor changes of the API

If the API change is not major, the things can be fixed in 10 minutes. I've written a new section about Fixing the type errors in the README of a validations-generator and added a link to FAQ of the project.

tl;dr:

1. Run `Library Classes` and `Full API` tests
1. See the warnings
1. Change the typings
1. Go back to step 1 until all problems are fixed.

### Major changes of the API

If changes are too big to patch the typings manually, we can always regenerate them. I've updated the instructions about how to do it.

The typings are not perfect, but using the tests it will be easy to fix them.

I believe it will take 1-3 hours of a middle developer, who sees this project for the first time, to generate new typings from scratch.

**Comment by goooseman** [comment]
One thing only is left.

I wanted to make TS generic type of `Hashtag`, `Location` and `User` classes to be auto-selected depending on `fullAPI` option passed to constructor.

I've tried to implement it the following way (dirty code WIP):

``` typescript
export interface IOptionsCommon {
//   ...types from IOptions without fullAPI
}

export type IOptionsRegularApi = IOptionsCommon & {
  // Extract the full amount of information from the API
  fullAPI?: false;
}

export type IOptionsFullApi = IOptionsCommon & {
  // Extract the full amount of information from the API
  fullAPI: true;
}

export type IOptions = IOptionsFullApi | IOptionsRegularApi;
```

So this way we have two different IOptions. Descriminator key is `fullAPI`.

``` typescript
export class Hashtag<O extends IOptions> extends Instagram<O extends IOptionsFullApi ? TFullApiPost : TPost> {
//...

export function hashtag(id, options: IOptions) {
  return new Hashtag<typeof options>(id, options).generator();
}
```

And in my **perfect world** it should work like this:
``` typescript
hashtag("2", {
  fullAPI: true,
})
```
should return `AsyncIterableIterator<TFullPost>`

And 
``` typescript
hashtag("2", {
  fullAPI: false,
})
```
should return `AsyncIterableIterator<TPost>`

But it just returns `AsyncIterableIterator<TPost>` in both cases.


Maybe you have an idea how to implement this correctly? 
I personally do not like this `<TPost | TFullApiPost>` typings, when it is obvious, that the second one will be used only when `fullAPI` option is true.
If not, then I'm ready to merge.

**Comment by ScriptSmith** [comment]
Great, thanks for your clarifications and insight into reacting to API changes.

I don't think what you're after is possible with just conditional types, but I think it could be with method overloading with the two IOptions types, I will investigate as well. I always look [here](https://www.typescriptlang.org/docs/handbook/advanced-types.html) when I'm trying to do something tricky with TS.

And my apologies, I should've phrased my comment more clearly. The full api mode itself doesn't really need a new test, I meant that it's return values should be tested as part of your strict mode tests.

File: src/api/instagram.ts [hunk 1]
```diff
@@ -254,8 +274,7 @@ export class Instagram {
           window["_sharedData"].entry_data.PostPage[0].graphql,
         );
       });
-      await this.addToPostBuffer(JSON.parse(data));
-
+      parsed = JSON.parse(data) as PostType;
       await postPage.close();
```
Comment (ScriptSmith) [inline src/api/instagram.ts]: The error catching I wrote here is probably too general to really be acceptable. However, if kept in its current form, postPage still isn't guaranteed to be closed because the error could occur when parsing the data, and then the postPage will not close and its reference will be lost.

I think of pages in puppeteer like malloc/free in C. You've got to keep track of them and not let them pile up.

Comment (goooseman) [reply src/api/instagram.ts]: Fixed with an additional try/catch block

**Comment by ScriptSmith** [comment]
I think something like [this](http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgPIAcwGED2BbPHEZAbwChlLkA3OAGwFcIBBALmQGcwpQBzAGgpVajCACF2IBngBG0MgF8yZUJFiIUGMAEZkEAB6QQAEw5pMuAkVJDKYAJ7oI7AEQgcUPPReLlq6PBI5mAATHqGECZmWpaExORUyA5OrjAMdHQ+SmTJKBDAYAAW0FrIALzBugA+wSEA3H7gARrIAAo4XNoJVO7GENrsXDwgAr4qTepB7Vwh3ZS9ECGD3HyC2WkgCGDA1uhQOEgcHAAUOJg7IBzsWtoAlOzTOg0bWxfIewcQR6fnRFe19zaHVCzwYm22u32hxOZwhl3Y+SKJUwgMe1SBMxsiQ4AHcCghCsgfnCOAA6XK3LGJKgIOAcFBuDxeTLsWzUxJQCBgBhQeJs9nshYDZCMzzefkCpQC5C0+kitIZFysCXsznc3lU6WCnB9Jby9KZFXUqVUJQKBpkBB-MDIdxiujld5Qr4nEgiJhsEVwFz8Gj0JgSZDaX25Vx25kuBS3BpWy42hUOiofaHHN3+liub2+93idjBpKOZz6xVRi2xjg4OgQUl0HC8Y4uACSIBg0E5xmQsIuVxct0tf0r1dr9fD9Gj-cug5rdeOCfHE64tqZ9C0-0RxSgpQqcz9ok9LizbJzgeDbNDItHhuy5fjBtXCIKG63mt3HszPqP6ZPgkS55cCayZRyynYcGwAFULDsuz+XsYwHKtp3rZMXWOS9V1ucdgIQ0DkO+BN0PHIA) should do what you're asking.

I'm not sure how well it works with io-ts, but it looks like it's correctly returning the correct type based on the mode set in the options interface.

**Comment by ScriptSmith** [comment]
In fact, we can use boolean literal types as well: [link](http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgPIAcwGED2BbPHEZAbwChlLkA3OAGwFcIBBALmQGcwpQBzAGgpVajCACF2IBngBG0MgF8yZUJFiIUGMAEZkEAB6QQAEw5pMuAkVJDKMBnTrMACgEl28OhwiLlq6PBI5mAATHqGECZmWpaExORUyPaOLu7I3Ey+ZGAAnugoEMBgABbQWsgAvMG6AD7BIQDcfuABGsjOOFzaCVQgOMYQ2uxcPCACWf7qQR1cIT2UfQMhw9x8gkpk9iAIYMDW6FA4SBwcABQ4mHsgHOxa2gCU7DM6TVs7V8gHRxAn55dEN3qj3anVCrwY212+0OxzOFyh13YhRKZUwwOetRBsxsiQ4AHcighisg-giOAA6ZJONz3HGJKgIODeJL0bzsWz0xJQCBgBhQeIczmcxaDdgAIj6UDw9DFgqFSiFyEZzIyEFYcs53N5-LpiuF-Qgy2QYqpsr1VAVFsUTTICABYGQkuldEqnxhPzOJBETDYxrgYv4NHoTAkyG0gapqQ8rIgCnuTTt1wdVNdX1hpy9wZY4v9ge94nY4aSDmpaVVcZtiY4ODoEHJdBwvFOYtcIBg0G5xmQ8KuNzF91tAJrdYbTad9Hjg+uw-rjdOVMnU64jpwUvoWkByNKUHKVXmQdEvrFuY5+dD4Y5kbc0a8Pg2VeTJY3SKK293uoPPpzAdPWfPgkSK8yygTJ7yHWtZybMUABU8ggLsewBfsE3Akc5zTD1TnHOgN3uScqxnUdTgw34qVwycgA)

**Comment by goooseman** [comment]
It is becoming a little bit complicated)

So I have created a factory for Instagram class. Any way to use a factory is not a bad idea IMO.

The cli.ts now looks a littttle bit cleaner, because the logic of choosing the right class has been moved to the factory method.

And it looks like working:
Screenshot from 2019-07-16 18-10-45
Screenshot from 2019-07-16 18-10-37

BUT

If I use through the function (hashtag, user, location and post ones, which are exported from index.ts), eveyrthing is screw up, and the return type is lost:
Screenshot from 2019-07-16 18-09-22


I've just pushed a WIP commit to show you the current state of the repo for today. 
I have 3 problems:

1. I'm not sure that you will accept the factory method
2. I'm not sure how I will fix the problem with hashtag, user, location and post functions
3. There are some type errors in cli.ts, which looks realy interesting. We are accessing some fields, which does not exist in my typings at all. This is weird. Maybe the API has been already changed and we did not know it?

Looking forward to your reply :)

**Comment by ScriptSmith** [comment]
I'm happy to use a factory pattern so long as the interface is obvious and coherent from the user's perspective. On my end the method signatures look to be derived correctly, but they're a little verbose. There might be some cleaning up we can do.

The type issues for `post.shortcode_media.edge_sidecar_to_children`, `post.shortcode_media.video_url` and others are valid and I don't think there's been an API change. The first two are types are for albums and videos respectively. I looked into your utils and saw  a couple of albums, so I'm not sure why they aren't being derived correctly. 

It should be getting types based on:
-  regular posts
- albums
- videos 

in 
- normal mode
- full mode

from 
- posts 
- tags
- locations
- users

**Comment by goooseman** [comment]
Of course, I can fix the typings easily, but first I want to get the typings error (some of the properties, which are not listed in the typings) in the tests.

Tests are the main typings-development utility, so if we have forgotten some of the properties, tests should fire a warning/throw as well.

Right now I'm investigating how to accomplish that.

File: src/api/instagram.ts (hunk 1 above)
Comment (goooseman) [reply src/api/instagram.ts]: In commit `12b4f1d8524a1765035a62ef1c31b8777e857b85` I've reverted back the change I've made in commit `fc74a790ba85abec4380509392bab3b0f19d05ab`.

I've double checked the code, which was before the commit:

If the error occured in parsing the data, code execution goes to catch block, where `postPage.close()` is.
If not - it will be closed (the code is on the next line).

If data validation will fail, the page is already closed. 

P.S. I've added two tests in strict mode section. They check, that error object message should contain validation information when validation fails. The code in commit `fc74a790ba85abec4380509392bab3b0f19d05ab` was throwing another error (can not close page which is closed), so I've written a regression page for this scenario.

**Comment by goooseman** [comment]
I've published a package for `io-ts` which exposes an `excess` type. This type check, that no additional properties were passed.

Continue to work on the project to fix all the typings and clean the factory method.

File: src/api/instagram.ts (hunk 1 above)
Comment (ScriptSmith) [reply src/api/instagram.ts]: You're right. I think I was being tripped up by the code folding in Github's web ui.

**Comment by goooseman** [comment]
Ok, types are fixed, types for the factory are written a little less verbose.

Everything looks nearly perfect right now.

The only problems we still have are `hashtag`, `location`, `user` functions which we export from `index.ts`.

Put this code in index.ts:

```
createApi("hashtag", "2", {
  fullAPI: true,
}).generator();

createApi("hashtag", "2", {
  fullAPI: false,
}).generator();

hashtag("2", {
  fullAPI: false,
});

hashtag("3", {
  fullAPI: true,
});
```

The problem is that the factory function returns the correct type.
But `hashtag` function, which just proxies the argument to the factory method loses the correct type (`TPost` or `TFullPost`) and shows the ouput as `TPost` always:
Screenshot from 2019-07-23 18-24-52


I do not know hot to fix it, but I assume, that we can deprecate `hashtag`, `location`, `user`, `post` function and tell users to use a new `createApi` factory method.

So we will be able some day to remove these functions in the next major (v2) release of the package.

**Comment by goooseman** [comment]
I believe it is ready to be merged

**Comment by ScriptSmith** [comment]
Great, I'll look over it once again in the next couple of days.

**Comment by ScriptSmith** [comment]
Merged! Thanks a bunch for all your feedback and code, I really appreciate it! :smiley:
